internal/oci/mutate/map.go

```diff
@@ -0,0 +1,167 @@
+var ErrSkipChildren = errors.New("skip child entities")
+
+// Map calls `fn` on the signed entity and each of its constituent entities (`SignedImageIndex`
+// or `SignedImage`) transitively.
+// Any errors returned by an `fn` are returned by `Map`.
+func Map(ctx context.Context, parent oci.SignedEntity, fn Mutator) (oci.SignedEntity, error) {
+	parent, err := fn(before(ctx), parent)
+	switch {
+	case errors.Is(err, ErrSkipChildren):
+		return parent, nil
+	case err != nil:
+		return nil, err
+	case parent == nil:
+		// If the function returns nil, it filters it.
+		return nil, nil
+	}
+
+	sii, ok := parent.(oci.SignedImageIndex)
+	if !ok {
+		return parent, nil
+	}
+	im, err := sii.IndexManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	// Track whether any of the child entities change.
+	changed := false
+
+	adds := []IndexAddendum{}
+	for _, desc := range im.Manifests {
+		switch desc.MediaType {
+		case types.OCIImageIndex, types.DockerManifestList:
+			x, err := sii.SignedImageIndex(desc.Digest)
+			if err != nil {
+				return nil, err
+			}
+
+			se, err := Map(ctx, x, fn)
+			if err != nil {
+				return nil, err
+			} else if se == nil {
+				// If the function returns nil, it filters it.
+				changed = true
+				continue
+			}
+
+			changed = changed || (x != se)
+			adds = append(adds, IndexAddendum{
+				Add: se.(oci.SignedImageIndex), // Must be an image index.
+				Descriptor: v1.Descriptor{
+					URLs:        desc.URLs,
+					MediaType:   desc.MediaType,
+					Annotations: desc.Annotations,
+					Platform:    desc.Platform,
+				},
+			})
+
+		case types.OCIManifestSchema1, types.DockerManifestSchema2:
+			x, err := sii.SignedImage(desc.Digest)
+			if err != nil {
+				return nil, err
+			}
+
+			se, err := fn(ctx, x)
+			if err != nil {
+				return nil, err
+			} else if se == nil {
+				// If the function returns nil, it filters it.
+				changed = true
+				continue
+			}
+
+			changed = changed || (x != se)
+			adds = append(adds, IndexAddendum{
+				Add: se.(oci.SignedImage), // Must be an image
+				Descriptor: v1.Descriptor{
+					URLs:        desc.URLs,
+					MediaType:   desc.MediaType,
+					Annotations: desc.Annotations,
+					Platform:    desc.Platform,
+				},
+			})
+
+		default:
+			return nil, fmt.Errorf("unknown mime type: %v", desc.MediaType)
+		}
+	}
+
+	if !changed {
+		return parent, nil
+	}
+
+	// Construct a new ImageIndex from the new consituent signed images.
+	result := AppendManifests(mutate.IndexMediaType(empty.Index, im.MediaType), adds...)
+
+	// Since the children changed, give the callback a crack at the new image index.
+	return fn(after(ctx), result)
+}
+
+// This is used to associate which pass of the Map a particular
+// callback is being invoked for.
+type mapPassKey struct{}
+
+// before decorates the context such that IsBeforeChildren(ctx) is true.
+func before(ctx context.Context) context.Context {
+	return context.WithValue(ctx, mapPassKey{}, "before")
```

Comment:
turn "before" and "after" into consts and/or enums?

Reply:
I was going to say "but I'll just call them before/after", but I can't because that's what the methods are called. 🙃 

Given the proximity of the functions, and the tests to make sure they agree, I'm sort of inclined to leave it 😅